snuba/query/processors/mapping_promoter.py

```diff
@@ -0,0 +1,137 @@
+    table_name: Optional[str]
+    # The nested column name
+    column_name: str
+    # The key column name in the mapping (like key for tags.key)
+    key_field: str
+    # The value column name in the mapping (like key for tags.value)
+    val_field: str
+    # The key found in the query (like key in tags[key])
+    key: str
+
+
+def match_subscriptable_reference(exp: Expression) -> Optional[SubscriptableMatch]:
+    """
+    Finds the expression, in the Clickhouse query, that loads the value
+    of a specific tag (or any nested column that represents a mapping,
+    like contexts).
+    It builds a SubscriptableMatch object to be used by the code that
+    processes such expression for convenience.
+
+    This is the shape of the expression matched here:
+    `arrayElement("tags.value", indexOf("tags.key", "myTag"))`
+    """
+
+    match = tag_pattern.match(exp)
+    if match is None:
+        return None
+
+    # TODO: There is should be a structured Column class (#963 - #966)
+    # to deal with references to nested columns instead of splitting
+    # the column name string.
+
+    # exp.parameters[0] is a Column with a name like `tags.value`
+    value_col_split = match.string(VALUE_COL_TAG_PARAM).split(".", 2)
+    # exp.parameters[1].parameters[0] is a Column with a name like `tags.key`
+    key_col_split = match.string(KEY_COL_TAG_PARAM).split(".", 2)
+
+    if len(value_col_split) != 2 or len(key_col_split) != 2:
+        return None
+
+    val_column, val_field = value_col_split
+    key_column, key_field = key_col_split
+
+    if val_column != key_column:
+        return None
+
+    table_name = match.scalar(TABLE_TAG_PARAM)
+    return SubscriptableMatch(
+        table_name=str(table_name) if table_name is not None else None,
+        column_name=val_column,
+        key_field=key_field,
+        val_field=val_field,
+        key=match.string(KEY_TAG_PARAM),
+    )
+
+
+class PromotedColumnsSpec(NamedTuple):
+    # The name of the key field in the mapping column
+    key_field: str
+    # The name of the value field in the mapping column
+    val_field: str
+    # The mapping between keys in the column and real columns
+    column_mapping: Mapping[str, str]
+
+
+class MappingColumnPromoter(QueryProcessor):
+    """
+    Promotes expressions that access the value of a mapping column by
+    replacing them with the corresponding promoted column provided in
+    the constructor.
+
+    Example: tags["myTag"]
+        -> arrayElement("tags.value", indexOf("tags.key", "myTag"))
+        -> toString(promoted_MyTag)
+
+    This happens if there is a promoted_MyTag column in the storage
+    that maps to the myTag tag.
+    """
+
+    def __init__(
+        self, columns: ColumnSet, mapping_spec: Mapping[str, PromotedColumnsSpec]
+    ) -> None:
+        # The ColumnSet of the dataset. Used to format promoted
+        # columns with the right type.
+        self.__columns = columns
```

Comment:
Isn't there a path to get these from the `Query` object?

Reply:
Actually there is. Did not think about it.